Angular-Electron Loading Error

## Issue Details

I'm working on an Angular-Electron-TypeORM application. I've forked this repo and updated the electron dependency version to v7. However, I've had an issue even using the official msnodesqlv8. When Angular builds it drops the bin directory. I've gotten around this by adding an asset configuration in the `angular.json` file:

```json
"assets": [
  {
    "glob": "**/*",
    "input": "./node_modules/msnodesqlv8/lib/bin",
    "output": "./bin"
  }
]
```

## Error Details

While the application loads in debug, it's still throwing the error: `Cannot find module` (stack below):

```
Error: Cannot find module 'C:\path\to\built\app\dist\build\Debug\sqlserverv8.node'
    at webpackEmptyContext (/C:/path/to/built/app/dist/main-es2015.4425b5d7dba874e9416e.js:171637)
    at Object.debugLoad (/C:/path/to/built/app/dist/main-es2015.4425b5d7dba874e9416e.js:171692)
    at Object../node_modules/msnodesqlv8/lib/bootstrap.js (/C:/path/to/built/app/dist/main-es2015.4425b5d7dba874e9416e.js:171734)
    at __webpack_require__ (/C:/path/to/built/app/dist/runtime-es2015.d6c52737d4587c65265f.js:80)
    at /C:/path/to/built/app/dist/main-es2015.4425b5d7dba874e9416e.js:171757
    at Object../node_modules/msnodesqlv8/lib/connection.js (/C:/path/to/built/app/dist/main-es2015.4425b5d7dba874e9416e.js:172193)
    at __webpack_require__ (/C:/path/to/built/app/dist/runtime-es2015.d6c52737d4587c65265f.js:80)
    at Object.<anonymous> (/C:/path/to/built/app/dist/main-es2015.4425b5d7dba874e9416e.js:174237)
    at Object../node_modules/msnodesqlv8/lib/sql.js (/C:/path/to/built/app/dist/main-es2015.4425b5d7dba874e9416e.js:174292)
    at __webpack_require__ (/C:/path/to/built/app/dist/runtime-es2015.d6c52737d4587c65265f.js:80)
```

If I try to use liveLoad, the error thrown is `None of the binaries loaded successfully. Is your node version >= 10.15 ?` (again, stack below)

```
Uncaught Error: None of the binaries loaded successfully. Is your node version >= 10.15 ?
    at failIfNoBinaryExported (main-es2015.1b472c6192034249cbbe.js:171719)
    at Object.liveLoad (main-es2015.1b472c6192034249cbbe.js:171706)
    at Object../node_modules/msnodesqlv8/lib/bootstrap.js (main-es2015.1b472c6192034249cbbe.js:171735)
    at __webpack_require__ (runtime-es2015.d6c52737d4587c65265f.js:80)
    at main-es2015.1b472c6192034249cbbe.js:171757
    at Object../node_modules/msnodesqlv8/lib/connection.js (main-es2015.1b472c6192034249cbbe.js:172193)
    at __webpack_require__ (runtime-es2015.d6c52737d4587c65265f.js:80)
    at Object.<anonymous> (main-es2015.1b472c6192034249cbbe.js:174237)
    at Object../node_modules/msnodesqlv8/lib/sql.js (main-es2015.1b472c6192034249cbbe.js:174292)
    at __webpack_require__ (runtime-es2015.d6c52737d4587c65265f.js:80)
```

I checked that the angular build directory contains the lib/bin directory and its contents. I also checked that it has a Build/Debug directory with the build contents. I did also verify security in the directory (just in case).

I feel as though I'm missing something simple. However, are there any considerations for using this library within an angular application? I think this may be worth a brief mention in the documentation if angular built apps have any special needs.

## Version Details

**Angular**: 8.2.12
**Electron**: 7.0.0
**Node**: 12.8.1
**msnodesqlv8**: 0.8.5 and forked with updated electron dependency

I attempted to add some additional debugging lines to the bootstrap.js file, and got the following (repeated for all .node drivers):

```
Attempting to load module: 'C:\path\to\app\dist\bin\sqlserverv8.node.v12.13.0.electron.v7.1.2.x64.node'...

Unable to load module: Cannot find module 'C:\path\to\app\dist\bin\sqlserverv8.node.v12.13.0.electron.v7.1.2.x64.node'
attemptToExportBinary @ main-es2015.709db0edd2e31fa9deb4.js:171716

Error: Cannot find module 'C:\path\to\app\dist\bin\sqlserverv8.node.v12.13.0.electron.v7.1.2.x64.node'
    at webpackEmptyContext (main-es2015.709db0edd2e31fa9deb4.js:171637)
    at attemptToExportBinary (main-es2015.709db0edd2e31fa9deb4.js:171713)
    at main-es2015.709db0edd2e31fa9deb4.js:171702
    at Array.forEach (<anonymous>)
    at Object.liveLoad (main-es2015.709db0edd2e31fa9deb4.js:171700)
    at Object../node_modules/msnodesqlv8/lib/bootstrap.js (main-es2015.709db0edd2e31fa9deb4.js:171739)
    at __webpack_require__ (runtime-es2015.d6c52737d4587c65265f.js:80)
    at main-es2015.709db0edd2e31fa9deb4.js:171761
    at Object../node_modules/msnodesqlv8/lib/connection.js (main-es2015.709db0edd2e31fa9deb4.js:172197)
    at __webpack_require__ (runtime-es2015.d6c52737d4587c65265f.js:80)
```

Again, I verified that the path is valid and that the node module referenced exists at the given location. However, for some reason, it's not able to import the app. I'm curious, you mentioned getting a test working on your side, did your tsconfig file align with what I have mine set to below? I'm curious if a build config is causing a problem.

## TS Config Details
**tsconfig.json**:
```json
  "compilerOptions": {
    "outDir": "./dist/out-tsc",
    "sourceMap": true,
    "declaration": false,
    "moduleResolution": "node",
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "target": "es6",
    "typeRoots": [
      "node_modules/@types"
    ],
    "lib": [
      "es2017",
      "es2016",
      "es2015",
      "dom"
    ]
  }
```
**tsconfig.app.json**:
```json
  "extends": "../tsconfig.json",
  "compilerOptions": {
    "outDir": "../out-tsc/app",
    "module": "esnext",
    "baseUrl": "",
    "types": []
  }
```

If we go back to basics and install the demo hello world electron app with its main js file and you add a using msnodesqlv8 with electron v7 does that app run and start without raising the error. If it does it shows we can at least run an electron app v7. If this works we then need to start looking at what stage the failure happens.  

Sent from my iPhone

> On 27 Nov 2019, at 00:33, Brad Van Fleet <notifications@github.com> wrote:
> 
> ﻿
> I attempted to add some additional debugging lines to the bootstrap.js file, and got the following (repeated for all .node drivers):
> 
> Attempting to load module: 'C:\path\to\app\dist\bin\sqlserverv8.node.v12.13.0.electron.v7.1.2.x64.node'...
> 
> Unable to load module: Cannot find module 'C:\path\to\app\dist\bin\sqlserverv8.node.v12.13.0.electron.v7.1.2.x64.node'
> attemptToExportBinary @ main-es2015.709db0edd2e31fa9deb4.js:171716
> 
> Error: Cannot find module 'C:\path\to\app\dist\bin\sqlserverv8.node.v12.13.0.electron.v7.1.2.x64.node'
>     at webpackEmptyContext (main-es2015.709db0edd2e31fa9deb4.js:171637)
>     at attemptToExportBinary (main-es2015.709db0edd2e31fa9deb4.js:171713)
>     at main-es2015.709db0edd2e31fa9deb4.js:171702
>     at Array.forEach (<anonymous>)
>     at Object.liveLoad (main-es2015.709db0edd2e31fa9deb4.js:171700)
>     at Object../node_modules/msnodesqlv8/lib/bootstrap.js (main-es2015.709db0edd2e31fa9deb4.js:171739)
>     at __webpack_require__ (runtime-es2015.d6c52737d4587c65265f.js:80)
>     at main-es2015.709db0edd2e31fa9deb4.js:171761
>     at Object../node_modules/msnodesqlv8/lib/connection.js (main-es2015.709db0edd2e31fa9deb4.js:172197)
>     at __webpack_require__ (runtime-es2015.d6c52737d4587c65265f.js:80)
> Again, I verified that the path is valid and that the node module referenced exists at the given location. However, for some reason, it's not able to import the app. I'm curious, you mentioned getting a test working on your side, did your tsconfig file align with what I have mine set to below? I'm curious if a build config is causing a problem.
> 
> TS Config Details
> 
> tsconfig.json:
> 
>   "compilerOptions": {
>     "outDir": "./dist/out-tsc",
>     "sourceMap": true,
>     "declaration": false,
>     "moduleResolution": "node",
>     "emitDecoratorMetadata": true,
>     "experimentalDecorators": true,
>     "target": "es6",
>     "typeRoots": [
>       "node_modules/@types"
>     ],
>     "lib": [
>       "es2017",
>       "es2016",
>       "es2015",
>       "dom"
>     ]
>   }
> tsconfig.app.json:
> 
>   "extends": "../tsconfig.json",
>   "compilerOptions": {
>     "outDir": "../out-tsc/app",
>     "module": "esnext",
>     "baseUrl": "",
>     "types": []
>   }
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.

--- a/README.md
+++ b/README.md
@@ -74,6 +74,19 @@ sql.query(connectionString, query, (err, rows) => {
 
 See our [TypeScript sample app](samples/typescript) for more details.
 
+### Webpack
+
+If you are using Webpack for your application, you need to:
+
+1. Add the [node-loader](https://webpack.js.org/loaders/node-loader/) as a dev dependency.
+2. Update your `webpack.config.js` to include the following under `module.rules`:
+    ```
+    {
+        test: /\.node$/,
+        use: 'node-loader'
+    }
+    ```
+
 ## Prepared Statements
 
 It is now possible to prepare one or more statements which can then be invoked

--- a/lib/bootstrap.js
+++ b/lib/bootstrap.js
@@ -29,7 +29,7 @@ const bootModule = ((() => {
 
   function debugLoad () {
     const binaryDir = __dirname
-    const filename = '..\\build\\Debug\\    try { module.exports = require('./bin/sqlserverv8.node'
+    const filename = '..\\build\\Debug\\sqlserverv8.node'
     console.log('DEBUG mode filename ' + filename)
     const native = path.join(binaryDir, filename)
     console.log('DEBUG loading from native ' + native)

--- a/lib/bootstrap.js
+++ b/lib/bootstrap.js
@@ -29,7 +29,7 @@ const bootModule = ((() => {
 
   function debugLoad () {
     const binaryDir = __dirname
-    const filename = '..\\build\\Debug\\sqlserverv8.node'
+    const filename = '..\\build\\Debug\\    try { module.exports = require('./bin/sqlserverv8.node'
     console.log('DEBUG mode filename ' + filename)
     const native = path.join(binaryDir, filename)
     console.log('DEBUG loading from native ' + native)
@@ -42,22 +42,24 @@ const bootModule = ((() => {
   }
 
   function liveLoad () {
-    files.forEach(file => {
-      if (noBinaryExported()) {
-        attemptToExportBinary(file)
-      }
-    })
-
-    failIfNoBinaryExported()
-
+    // Try loading the correct lib for this platform/environment 
     // If the require fails, we can just ignore the exception and continue trying
-    function attemptToExportBinary (filename) {
-      try {
-        const native = path.join(binaryDir, filename)
-        module.exports = require(native)
-      } catch (e) {
-      }
-    }
+    try { module.exports = require('./bin/sqlserverv8.node.v10.16.0.ia32.node'); } catch (e) {};    
+    try { module.exports = require('./bin/sqlserverv8.node.v10.16.0.x64.node'); } catch (e) {};
+    try { module.exports = require('./bin/sqlserverv8.node.v11.15.0.electron.v5.0.12.ia32.node'); } catch (e) {};
+    try { module.exports = require('./bin/sqlserverv8.node.v11.15.0.electron.v5.0.12.x64.node'); } catch (e) {};
+    try { module.exports = require('./bin/sqlserverv8.node.v11.15.0.ia32.node'); } catch (e) {};
+    try { module.exports = require('./bin/sqlserverv8.node.v11.15.0.x64.node'); } catch (e) {};
+    try { module.exports = require('./bin/sqlserverv8.node.v12.13.0.electron.v6.1.5.ia32.node'); } catch (e) {};
+    try { module.exports = require('./bin/sqlserverv8.node.v12.13.0.electron.v6.1.5.x64.node'); } catch (e) {};
+    try { module.exports = require('./bin/sqlserverv8.node.v12.13.0.electron.v7.1.2.ia32.node'); } catch (e) {};
+    try { module.exports = require('./bin/sqlserverv8.node.v12.13.0.electron.v7.1.2.x64.node'); } catch (e) {};
+    try { module.exports = require('./bin/sqlserverv8.node.v12.13.1.ia32.node'); } catch (e) {};
+    try { module.exports = require('./bin/sqlserverv8.node.v12.13.1.x64.node'); } catch (e) {};
+    try { module.exports = require('./bin/sqlserverv8.node.v13.2.0.ia32.node'); } catch (e) {};
+    try { module.exports = require('./bin/sqlserverv8.node.v13.2.0.x64.node'); } catch (e) {};
+
+    failIfNoBinaryExported();
 
     function failIfNoBinaryExported () {
       if (noBinaryExported()) {

--- a/lib/bootstrap.js
+++ b/lib/bootstrap.js
@@ -23,11 +23,9 @@
 // ---------------------------------------------------------------------------------------------------------------------------------
 
 const bootModule = ((() => {
-  const path = require('path')
-  const binaryDir = path.join(__dirname, 'bin')
-  const files = require('fs').readdirSync(binaryDir)
-
+  
   function debugLoad () {
+    const path = require('path')
     const binaryDir = __dirname
     const filename = '..\\build\\Debug\\sqlserverv8.node'
     console.log('DEBUG mode filename ' + filename)

